alternator/ttl.cc

```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
 
+// expiration_service is a sharded service responsible for cleaning up expired
+// items in all tables with per-item expiration enabled. Currently, this means
+// Alternator tables with TTL configured via a UpdateTimeToLeave request.
+//
+// Here is a brief overview of how the expiration service works:
+//
+// An expiration thread on each shard periodically scans the items (i.e.,
+// rows) owned by this shard, looking for items whose chosen expiration-time
+// attribute indicates they are expired, and deletes those items.
+// The expiration-time "attribute" can be either an actual Scylla column
+// (must be numeric) or an Alternator "attribute" - i.e., an element in
+// the ATTRS_COLUMN_NAME map<utf8,bytes> column where the numeric expiration
+// time is encoded in DynamoDB's JSON encoding inside the bytes value.
+// To avoid scanning the same items RF times in RF replicas, only one node is
+// responsible for scanning a token range at a time. Normally, this is the
+// node owning this range as a "primary range" (the first node in the ring
+// with this range), but when this node is down, other nodes may take over
+// (FIXME: this is not implemented yet).
+// An expiration thread is reponsible for all tables which need expiration
+// scans. FIXME: explain how this is done with multiple tables - parallel,
+// staggered, or what?
+// The expiration thread scans item using CL=QUORUM to ensures that it reads
+// a consistent expiration-time attribute. This means that the items are read
+// locally and in addition QUORUM-1 additional nodes (one additional node
+// when RF=3) need to read the data and send digests.
+// FIXME: explain if we can read the exact attribute or the entire map.
+// When the expiration thread decides that an item has expired and wants
+// to delete it, it does it using a CL=QUORUM write. This allows this
+// deletion to be visible for consistent (quorum) reads. The deletion,
+// like user deletions, will also appear on the CDC log and therefore
+// Alternator Streams if enabled (FIXME: explain how we mark the
+// deletion different from user deletes. We don't do it yet.).
+expiration_service::expiration_service(database& db, service::storage_proxy& proxy)
+        : _db(db)
+        , _proxy(proxy)
+{
+    //FIXME: add metrics for the service
+    //setup_metrics();
+}
+
+// The following is_expired() functions all check if an item with the given
+// expiration time has expired, according to the DynamoDB API rules.
+// The rules are:
+// 1. If the expiration time attribute's value is not a number type,
+//    the item is not expired.
+// 2. The expiration time is measured in seconds since the UNIX epoch.
+// 3. If the expiration time is more than 5 years in the past, it is assumed
+//    to be malformed and ignored - and the item does not expire.
+static bool is_expired(gc_clock::time_point expiration_time, gc_clock::time_point now) {
+    constexpr gc_clock::duration five_years = std::chrono::seconds(5 * 365 * 24 * 3600);
```

Comment:
You can write std::chrono::years(5);

Reply:
I actually tried that first - and it doesn't work. After spending 5 minutes on it I gave up and did other more important things.
My guess is that a "year" can't be translated to a fixed number of seconds because it depends on leap years and so on, so they didn't want to allow such conversion.

Reply:
You can go with `std::chrono::days` or `std::chrono::weeks` (: months can probably be tricky too, but we certainly don't have leap days. That problem seems weird though, according to https://en.cppreference.com/w/cpp/chrono/duration, std::chrono::years is a very specific ratio and does not consider leap years to be a problem. But you're right that this thing is probably not important enough to spend a considerable amount of time debugging it.

Reply:
I wonder if I just mispelled "years" and wrote "year" - which is a yet another class (https://en.cppreference.com/w/cpp/chrono/year) which is completely different and wouldn't work.
I'll try to use std::chrono::years, I think it's clearest and it doesn't matter if it covers another extra 5/4th of a day or something.

Reply:
So, std::chrono::years did work and I'll switch to it instead.
I am guessing the reason it didn't work previously was that I mistyped std::chrono::year, which also exists but is something else (and not convertable to seconds, obviously).